File: tools/check-swish.pl
:- encoding(utf8).

% Asks exactly what SWISH asks before running a query: whether the goal, and
% everything that goal calls, is allowed by the sandbox.
%
% Driven by tools/check-swish.py, one process per example:
%
%     swipl -g "check('ejemplos/capitulo-01/familia.pl', [\"abuelo(juan, X)\"])" \
%           -t halt tools/check-swish.pl

:- use_module(library(sandbox)).
:- use_module(library(pengines_io), [pengine_io_goal_expansion/2]).

check(File, Queries) :-
    consult(File),
    findall(Query-Reason,
            ( member(Query, Queries), rejected(Query, Reason) ),
            Rejected),
    forall(member(Query-Reason, Rejected),
           format("  ~w~n      ~p~n", [Query, Reason])),
    Rejected == [].

% A query is rejected when safe_goal/1 throws (the usual case: the sandbox names
% the forbidden predicate) or when it fails without saying anything.
rejected(Query, Reason) :-
    term_string(Goal, Query),
    (   catch(safe_goal(Goal), Error, true)
    ->  nonvar(Error),
        Reason = Error
    ;   Reason = 'safe_goal/1 failed'
    ).

% SWISH does not run the program against the plain built-ins. It loads
% library(pengines_io), which replaces write/1, nl/0, tab/1, read/1, listing/1
% and friends with pengine_write/1, pengine_nl/0 and so on, so their output
% reaches the answer pane instead of a stream, and declares those replacements
% sandbox-safe.
%
% Here there is no pengine, so write/1 resolves to system:write/1, which the
% sandbox rightly refuses: it writes to a stream. Without the declarations
% below, every example that prints anything would be reported as impossible in
% SWISH, which is the opposite of the truth.
%
% So the predicates SWISH swaps out are declared safe. The list is spelled out
% one by one because library(sandbox) verifies each declaration at load time and
% refuses a rule with a variable head. It is the list of pengine_io_predicate/1
% in library/ext/pengines/pengines_io.pl, and `mismatch/1` below fails the load
% if the two ever drift apart.
%
% format/1,2 are in that list but need no declaration: library(sandbox) already
% allows them through safe_meta/2, which parses the format string and checks
% whatever it would call. Declaring them safe outright would be weaker than what
% Loaded so the declarations below can name the module each predicate really
% lives in: library(sandbox) refuses to declare a predicate it cannot see.
:- use_module(library(listing)).
:- use_module(library(readutil)).

sandbox:safe_primitive(system:write(_)).
sandbox:safe_primitive(system:writeln(_)).
sandbox:safe_primitive(system:writeq(_)).
sandbox:safe_primitive(system:write_canonical(_)).
sandbox:safe_primitive(system:write_term(_, _)).
sandbox:safe_primitive(system:print(_)).
sandbox:safe_primitive(system:nl).
sandbox:safe_primitive(system:tab(_)).
sandbox:safe_primitive(system:flush_output).
sandbox:safe_primitive(system:read(_)).
sandbox:safe_primitive(prolog_listing:listing).
sandbox:safe_primitive(read_util:read_line_to_string(_, _)).
sandbox:safe_primitive(read_util:read_line_to_codes(_, _)).

% The tracer. These are not declared by any library shipped with SWI-Prolog:
% they come from SWISH's own lib/trace.pl, which lives in the SWISH sources
% (https://github.com/SWI-Prolog/swish), so nothing installed locally can tell
% us they are allowed.
%
% Verified against the real service on 2026-09-20 by asking its pengines API to
% run `trace, abuelo(juan, X)` over a three-clause program: it answered
% "event":"success" with X = luis. The same request with shell/1 came back
% "permission_error", which is what makes that a meaningful test rather than an
% endpoint that runs anything.
%
% This matters for chapter 1, whose section on watching Prolog search is built
% on trace/0.
sandbox:safe_primitive(system:trace).
sandbox:safe_primitive(system:notrace).
sandbox:safe_primitive(system:tracing).
sandbox:safe_primitive(system:deterministic(_)).

% listing/1 and portray_clause/1 are meta-predicates, and safe_primitive/1
% refuses those by design. safe_meta/2 is the hook for them: it states which
% goals the meta-predicate will call, and these two call none.
sandbox:safe_meta(prolog_listing:listing(_), []).
sandbox:safe_meta(prolog_listing:portray_clause(_), []).

% Three of the predicates SWISH substitutes are left undeclared on purpose.
deliberately_undeclared(format(_),
    'library(sandbox) already allows it through safe_meta/2, which parses the format string').
deliberately_undeclared(format(_, _),
    'library(sandbox) already allows it through safe_meta/2, which parses the format string').
deliberately_undeclared(display(_),
    'an obsolete Edinburgh predicate; this book never uses it, so it stays out').

undeclared_substitution(Goal) :-
    pengine_io_goal_expansion(Goal, _),
    \+ deliberately_undeclared(Goal, _),
    \+ catch(sandbox:safe_primitive(_:Goal), _, fail),
    \+ catch(sandbox:safe_meta(_:Goal, _), _, fail).

% Warns at load time if SWISH substitutes a predicate this file neither declares
% nor deliberately leaves out, so a new SWI-Prolog release cannot quietly widen
% Stops the load if SWISH substitutes a predicate this file neither declares nor
% deliberately leaves out, so a new SWI-Prolog release cannot quietly widen the
% gap between what SWISH runs and what this checker allows. A checker that has
% fallen out of step must not report success. It goes last: a directive runs
% while the file loads, so the clauses it calls have to be in already.
:- (   findall(Goal, undeclared_substitution(Goal), Goals),
       Goals \== []
   ->  forall(member(Goal, Goals),
              format(user_error,
                     "ERROR: check-swish.pl neither declares nor excludes ~q, which SWISH replaces~n",
                     [Goal])),
       halt(1)
   ;   true
   ).

File: tools/check-swish.py
#!/usr/bin/env python
"""Check that every example can run in SWISH, without opening SWISH.

SWISH does not run just anything: before running a query it asks
`library(sandbox)` whether the goal, and everything that goal calls, is allowed.
An example that uses `shell/1`, opens a file or starts a thread reaches the book
with a broken link, and nobody finds out until a student clicks it.

This asks the same question, through the same `library(sandbox)`, about every
`%?-` query in the example's header.

An example that cannot run in SWISH (files, threads, user interfaces,
compilation: most of part III) says so in its header and is skipped:

    % solo-local: abre un archivo del disco, que el sandbox de SWISH no permite.

The text then prints "Solo local (swipl)" where the link would go.

    ./tools/check-swish.py                 all of them
    ./tools/check-swish.py capitulo-07     one chapter

Exits 1 if an example that promises SWISH does not pass the sandbox.
"""

import json
import shutil
import subprocess
import sys
from pathlib import Path

import examples

DRIVER = Path(__file__).resolve().parent / 'check-swish.pl'
LIMIT = 60


def goal(example):
    # A Prolog string takes the same escapes as a JSON one, so json.dumps leaves
    # each query ready to paste into the goal.
    queries = ','.join(json.dumps(q.rstrip('.')) for q in example.queries)
    return f"check('{example.path.as_posix()}',[{queries}])"


def check(swipl, example):
    if not examples.declares_encoding(example.source):
        return False, 'is missing its first line,  :- encoding(utf8).'
    if example.local_only:
        return True, f'local only: {example.local_only}'
    if not example.queries:
        # With no query there is no link to test, and the reader would not know
        # what to ask it.
        return False, 'declares no %?- query in its header'
    try:
        done = subprocess.run(  # noqa: S603
            [swipl, '-g', goal(example), '-t', 'halt', str(DRIVER)],
            capture_output=True,
            text=True,
            errors='replace',
            timeout=LIMIT,
        )
    except subprocess.TimeoutExpired:
        return False, f'the sandbox did not answer in {LIMIT} seconds'
    output = (done.stdout + done.stderr).strip()
    head = '\n'.join(output.splitlines()[:10])
    # swipl exits 0 even when a declaration inside check-swish.pl was refused at
    # load time, which would quietly turn this check into a no-op. So any ERROR:
    # in the output is a failure, whatever the exit status says.
    if done.returncode == 0 and 'ERROR:' in output:
        return False, 'check-swish.pl did not load cleanly:\n' + head
    if done.returncode != 0:
        return False, 'the SWISH sandbox rejects it:\n' + head
    return True, f'{len(example.queries)} query(ies) pass the sandbox'


def main():
    swipl = shutil.which('swipl')
    if swipl is None:
        print('swipl is not on the PATH.', file=sys.stderr)
        return 1
    wanted = [a for a in sys.argv[1:] if not a.startswith('-')]
    found = examples.all_examples(wanted)
    if not found:
        print('No examples to check%s' % (' by that name' if wanted else ''), file=sys.stderr)
        return 1
    width = max(len(e.relative) for e in found)
    failed = []
    for example in found:
        ok, detail = check(swipl, example)
        shown = detail if ok else 'FAILS'
        print(f'{example.relative:<{width}} {shown}')
        if not ok:
            print('    {}'.format(detail.replace('\n', '\n    ')))
            failed.append(example.relative)
    print('All of them open in SWISH.' if not failed else 'Failed: {}'.format(', '.join(failed)))
    return 1 if failed else 0


if __name__ == '__main__':
    sys.exit(main())
